### PR TITLE
Improve Azure deployment stability

### PR DESCRIPTION
**Pull request by Chaffelson**
Dependent on:
https://github.com/cloudera-labs/cdpy/pull/26
https://github.com/cloudera-labs/cloudera.cloud/pull/24

Add explicit test for Azure Storage Account being unavailable for use in this deployment
remove duplicate usage of __azure_netapp_vol_info as variable
Handle edge cases for preparation of Netapp NFS Mount when deploying ML automatically on Azure
Introduce wait and retry controls for handling Azure eventual consistency when negotiating between Ansible Controller, Azure Control Plane, and CDP Control Plane
Move default Azure minimal policy json from private gist to cloudera-labs snippets
Correct Azure App name where sometimes referred to with http:// header and sometimes not, resulting in idempotence failures. Oops.
Introduce more robust validations that Service principals and other objects created by az CLI are populated as expected
Ensure that Azure objects are consistently bound to the Azure namespace created from the name_prefix
Provide more user friendly errors when Azure App and Service Principal creation doesn't go as planned
Add tunnel and public endpoint control support to Azure Environment creation in line with AWS offering
Fix ML submission preparation to include nfs information following existing combination patterns
Swap order of Runtime initialization tasks to handle provider-specific tasks before general tasks, to allow Azure-specific values to be populated. No impact on other implementations
Explicitly derive Azure NFS Mount information in Runtime deployment from earlier Infrastructure deployment steps
Fix purge teardown edgecase where child services are not deleted if at least one child service is not present in the Definition. Now they are removed if purge is set and any are found regardless of provided Definition plan.

Meta-changelist:
Fix formatting in cloud.datahub_template_info
Add retry to address intermittent failure in datahub_template_info listing of datahub templates in CDP 7.2.10
cdpy now recognises that CML deployments may fail during provisioning and responds accordingly
cdpy will now automatically retry Azure cross-account credential creation in CDP when receiving eventual consistency errors

Signed-off-by: Daniel Chaffelson <chaffelson@gmail.com>

### CODE REVIEW

File: roles/infrastructure/tasks/setup_azure_storage.yml [hunk 1]
```diff
@@ -32,6 +34,23 @@
       location:  "{{ infra__region }}"
       tags: "{{ infra__tags }}"
 
+- name: Check Storage Account creation result for failure
+  when:
+    - __infra_az_stor_acccount_test.rc is defined
+    - __infra_az_stor_acccount_test.rc == 1
+  block:
+    - name: Test Azure Storage Account creation for StorageAccountAlreadyTaken
```
Comment (cmperro) [inline roles/infrastructure/tasks/setup_azure_storage.yml:42]: We should do this for AWS too. Many a time I have torn down without deleting data. And I get a strange AWS error. 

I realize I just volunteered to do this...

File: roles/infrastructure/tasks/setup_azure_storage.yml [hunk 2]
```diff
@@ -90,6 +109,13 @@
...
         size: "{{ infra__azure_netapp_vol_size }}"
 
-    - name: Prepare netapp vol info for submission
+    - name: Prepare netapp vol info for submission during initial creation
+      when: __azure_netapp_vol_details.msg is defined
+      set_fact:
+        infra__azure_nfs_mount: "{{ __azure_netapp_vol_details.msg }}"
+
+    # Yay for Azure consistency
+    - name: Prepare netapp vol info for submission during recreation
```
Comment (cmperro) [inline roles/infrastructure/tasks/setup_azure_storage.yml:118]: This can take a long time (3-6 minutes). It's MORE than a fair assumption to say that it will be done by the time the CDP environment is up. Just wondering out loud if there's any benefit to waiting for it.

File: roles/platform/tasks/initialize_azure.yml [hunk 3]
```diff
@@ -64,10 +64,27 @@
...
       ansible.builtin.set_fact:
         plat__azure_application_service_principal_objuuid: "{{ __azure_application_service_principals_list.stdout | from_json | community.general.json_query('[0].objectId') }}"
 
+    - name: Check that Azure Service Principal ID is now set
+      ansible.builtin.assert:
+        quiet: yes
+        that: plat__azure_application_service_principal_objuuid | length > 0
+        fail_msg: "Azure Service Principal Object ID appears to be length 0, please check and try again"
+
+- name: Print current plat__azure_application_service_principal_objuuid
```
Comment (cmperro) [inline roles/platform/tasks/initialize_azure.yml:84]: Did you mean to leave this in?

File: roles/platform/tasks/setup_azure_authz.yml [hunk 4]
```diff
@@ -25,36 +25,66 @@
...
+      block:
+        - name: Issue Azure App Delete command
+          command: "az ad app delete --id {{ plat__azure_xaccount_app_uuid }}"
+
+        - name: Wait for consistency
+          ansible.builtin.pause:
+            seconds: "{{ plat__azure_consistency_wait }}"
 
-    # Owner role is required for DWX
+    # Owner role is required for DWX if you are thinking of modifying this task
```
Comment (cmperro) [inline roles/platform/tasks/setup_azure_authz.yml:36]: Not anymore... I realize it's outside the scope of this PR, but DWX _just_ needs Contributor these days.

Comment (cmperro) [reply roles/platform/tasks/setup_azure_authz.yml:36]: Oooph. And single resource group. I'll create "Issues".

File: roles/runtime/tasks/initialize_base.yml [hunk 5]
```diff
@@ -148,7 +148,8 @@
         overlay_network: "{{ __ml_config.network | default(include.network) | default({}) }}"
         base_instance_group: "{{ run__ml_k8s_request_base }}"
         config:
-          name: "{{ __ml_config.name | default([run__namespace, __ml_config.suffix | default(include.suffix) | default(run__ml_suffix)] | join('-')) }}"
+          name: "{{ __ml_config.name | default([run__namespace_cdp, __ml_config.suffix | default(include.suffix) | default(run__ml_suffix)] | join('-')) }}"
+          nfs: "{{ __ml_config.nfs | default(run__azure_nfs_mount | default(omit)) }}"
```
Comment (cmperro) [inline roles/runtime/tasks/initialize_base.yml:152]: Why have the default(omit)? Wouldn't it be better to fail here than get all the way to the cloud module that tries to create the CML workspace?

File: roles/runtime/tasks/setup_base.yml [hunk 6]
```diff
@@ -46,10 +46,10 @@
     governance: "{{ __ml_config_item.raw.governance | default(omit) }}"
     metrics: "{{ __ml_config_item.raw.metrics | default(omit) }}"
     database: "{{ __ml_config_item.raw.database | default(omit) }}"
-    nfs: "{{ __ml_config_item.raw.nfs | default(omit) }}"
+    nfs: "{{ __ml_config_item.nfs | default(omit) }}"
     nfs_version: "{{ __ml_config_item.raw.nfs_version | default(omit) }}"
     ip_addresses: "{{ __ml_config_item.raw.ip_addresses | default(omit) }}"
-    public_loadbalancer: "{{ __ml_config_item.raw.public_loadbalancer | default(run__public_endpoint_access) | default(omit) }}"
+    public_loadbalancer: "{{ __ml_config_item.raw.public_loadbalancer | default(omit) }}"
```
Comment (cmperro) [inline roles/runtime/tasks/setup_base.yml:52]: Waiting to do all the experiences at once? Re: removing the run__public_endpoint_access?

File: roles/infrastructure/tasks/setup_azure_storage.yml (hunk 1 above)
Comment (Chaffelson) [reply roles/infrastructure/tasks/setup_azure_storage.yml:42]: @wmudge suggested we put this into the AWS side of things, and even suggested the approach I have used here, so it would be good to cross pollenate as a pattern where we know a common failure happens and we want to give a very specific instruction to the user to avoid confusion.

File: roles/infrastructure/tasks/setup_azure_storage.yml (hunk 2 above)
Comment (Chaffelson) [reply roles/infrastructure/tasks/setup_azure_storage.yml:118]: Because it fails in weird ways, especially due to backend Netapp actions that sometimes just fail and require an Azure support ticket to resolve. Therefore the user may get a weird CML deployment failure when actually their NFS is the problem.

File: roles/runtime/tasks/initialize_base.yml (hunk 5 above)
Comment (Chaffelson) [reply roles/runtime/tasks/initialize_base.yml:152]: because I am being overly clever by reusing the same preparation task for AWS and Azure, so if you are doing an AWS preparation and thus do not supply the NFS information, they configuration item will be omitted, and if you are doing Azure and it is missing you'll get the CLI failure. We could add the explicit requirement to cdpy to prevent it from submitting to the CLI tho, which would be nice.

File: roles/runtime/tasks/setup_base.yml (hunk 6 above)
Comment (Chaffelson) [reply roles/runtime/tasks/setup_base.yml:52]: The default is already set in the ML preparation step, and the user can override by setting it in their custom submission, so this was simply extraneous.

File: roles/platform/tasks/initialize_azure.yml (hunk 3 above)
Comment (Chaffelson) [reply roles/platform/tasks/initialize_azure.yml:84]: good spot, removing

**Review by wmudge — APPROVED** [review]
